core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -260,20 +261,20 @@ class FileMessageSet private[kafka](@volatile var file: File,
         sizeOffsetBuffer.rewind()
         val offset = sizeOffsetBuffer.getLong()
         val size = sizeOffsetBuffer.getInt()
-        if(size < Message.MinMessageOverhead)
+        if(size < Message.MinMessageOverhead || location + SizeOffsetLength + size > end)
```

Comment:
"end" is not inclusive. As long as we don't need the byte at "end" or beyond, we can stay in the iterator. So, in this case, > seems correct.